xls/examples/hack.x

```diff
@@ -0,0 +1,258 @@
+}
+
+fn alu(x: u16, y:u16, c:u6) -> (u16, u1, u1) {
+    let (zx, nx, zy, ny, f, no) = (c[5+:u1], c[4+:u1], c[3+:u1], c[2+:u1], c[1+:u1], c[0+:u1]);
+    let x' = z_bit(zx, x);
+    let x'' = n_bit(nx, x');
+    let y' = z_bit(zy, y);
+    let y'' = n_bit(ny, y');
+    let sum: u16 = (x'' + y'');
+    let and: u16 = x'' & y'';
+    let out: u16 = sum if f else and;
+    let out': u16 = !out if no else out;
+    let zr = (out' == u16:0);
+    let ng = (out' as s16 < s16:0);
+    (out', zr, ng)
+}
+
+#![test]
+fn alu_test() {
+    let (out, zr, ng) = alu(u16:2, u16:8, COMP_D_PLUS_1[0:6]);
+    let _ = assert_eq(out, u16:3);
+    let _ = assert_eq(zr, u1:0);
+    let _ = assert_eq(ng, u1:0);
+    let (out, zr, ng) = alu(u16:2, u16:8, COMP_D_AND_A[0:6]);
+    let _ = assert_eq(out, u16:0);
+    let _ = assert_eq(zr, u1:1);
+    let _ = assert_eq(ng, u1:0);
+    let (out, zr, ng) = alu(u16:8, u16:2, COMP_A_MINUS_D[0:6]);
+    let _ = assert_eq(out, s16:-6 as u16);
+    let _ = assert_eq(zr, u1:0);
+    let _ = assert_eq(ng, u1:1);
+    _
+}
+
+fn run_c_instruction(pc: u16, ins: u16, rd: u16, ra: u16, rm: u16) -> (u16, u16, u16, u16, u1) {
+    let (comp, dest, jump) = decode_c_instruction(ins);
+    let x = rd;
+    let a = comp[6+:u1];
+    let y = rm if a else ra;
+    let (out, zr, ng) = alu(x, y, comp[0+:u6]);
+    let rd' = out if (dest & DEST_D) == DEST_D else rd;
+    let ra' = out if (dest & DEST_A) == DEST_A else ra;
+    let (rm', wm) = (out, u1:1) if (dest & DEST_M) == DEST_M else (rm, u1:0);
+    let flags: u3 = ng ++ zr ++ !ng;
+    let pc' = ra' if (jump & flags) != u3:0 else pc + u16:1;
+    (pc', rd', ra', rm', wm)
+}
+
+#![test]
+fn run_c_test() {
+    // MD=D+1
+    let (pc, rd, ra, rm, wm) = run_c_instruction(u16:0, u16:0b111_0011111_011_000, u16:1, u16:2, u16:3);
+    let _ = assert_eq(pc, u16:1);
+    let _ = assert_eq(rd, u16:2);
+    let _ = assert_eq(ra, u16:2);
+    let _ = assert_eq(rm, u16:2);
+    // D+1;JLE
+    let (pc, rd, ra, rm, wm) = run_c_instruction(u16:0, u16:0b111_0011111_000_110, s16:-1 as u16, u16:2, u16:3);
+    let _ = assert_eq(pc, u16:2);
+    let _ = assert_eq(rd, s16:-1 as u16);
+    let _ = assert_eq(ra, u16:2);
+    let _ = assert_eq(rm, u16:3);
+    // 0;JMP
+    let (pc, rd, ra, rm, wm) = run_c_instruction(u16:0, u16:0b111_0101010_000_111, u16:1, u16:2, u16:3);
+    let _ = assert_eq(pc, u16:2);
+    let _ = assert_eq(rd, u16:1);
+    let _ = assert_eq(ra, u16:2);
+    let _ = assert_eq(rm, u16:3);
+    _
+}
+
+
+fn cpu(pc: u16, rd: u16, ra: u16, ram: u16[32], rom: u16[32]) -> (u16, u16, u16, u16[32])  {
+    let ins = rom[pc];
+    let rm = ram[ra];
+    let (pc', rd', ra', rm', wm) = match ins[15+:u1] {
+        u1:0 => run_a_instruction(pc, ins, rd, ra, rm),
+        u1:1 => run_c_instruction(pc, ins, rd, ra, rm),
+        _ => fail!((pc, rd, ra, rm, u1:0)),
+    };
+    (pc', rd', ra', update(ram, ra', rm') if wm else ram)
+}    
+
+#![test]
+fn run_cpu() {
+    let rom = u16[32]:[
```

Comment:
Can you add an encode instruction instruction function? It would help code readability.

The function can concat the inputs and reference the constants declared earlier.

Reply:
Done